File: desktop/macos/Desktop/Sources/Bluetooth/Session/BluetoothConnectionLease.swift
@preconcurrency import CoreBluetooth
import Foundation

/// Manager-owned identity for one CoreBluetooth connection attempt.
///
/// CoreBluetooth delegate callbacks carry only a peripheral. The manager keeps
/// a lease active until a terminal delegate callback (or central reset), so a
/// later session cannot reuse that peripheral identity while an old callback is
/// still possible.
struct BluetoothConnectionLease: Equatable, Hashable, Sendable {
  let peripheralID: UUID
  let token: UInt64
  let sessionGeneration: UInt64
}

enum BluetoothConnectionLeaseError: LocalizedError, Equatable {
  case leaseAlreadyActive(BluetoothConnectionLease)
  case centralUnavailable

  var errorDescription: String? {
    switch self {
    case .leaseAlreadyActive:
      return "A previous Bluetooth connection attempt is still draining"
    case .centralUnavailable:
      return "Bluetooth is not powered on"
    }
  }
}

@MainActor
protocol BluetoothCentralConnectionControlling: AnyObject {
  func beginConnection(
    to peripheral: CBPeripheral,
    sessionGeneration: UInt64
  ) throws -> BluetoothConnectionLease

  func cancelConnection(
    to peripheral: CBPeripheral,
    lease: BluetoothConnectionLease
  )
}

@MainActor
final class BluetoothConnectionLeaseRegistry {
  private enum Phase: Equatable {
    case connecting
    case connected
    case cancelling
  }

  private struct Entry {
    let lease: BluetoothConnectionLease
    var phase: Phase
  }

  private var nextToken: UInt64 = 0
  private var entries: [UUID: Entry] = [:]

  func begin(
    peripheralID: UUID,
    sessionGeneration: UInt64
  ) throws -> BluetoothConnectionLease {
    if let existing = entries[peripheralID]?.lease {
      throw BluetoothConnectionLeaseError.leaseAlreadyActive(existing)
    }

    nextToken &+= 1
    let lease = BluetoothConnectionLease(
      peripheralID: peripheralID,
      token: nextToken,
      sessionGeneration: sessionGeneration
    )
    entries[peripheralID] = Entry(lease: lease, phase: .connecting)
    return lease
  }

  /// Marks cancellation without releasing identity. The lease remains fenced
  /// until CoreBluetooth reports failure/disconnect or the central resets.
  @discardableResult
  func requestCancellation(_ lease: BluetoothConnectionLease) -> Bool {
    guard var entry = entries[lease.peripheralID], entry.lease == lease else {
      return false
    }
    entry.phase = .cancelling
    entries[lease.peripheralID] = entry
    return true
  }

  func markConnected(
    peripheralID: UUID
  ) -> (lease: BluetoothConnectionLease, shouldCancel: Bool)? {
    guard var entry = entries[peripheralID] else { return nil }
    let shouldCancel = entry.phase == .cancelling
    if !shouldCancel {
      entry.phase = .connected
      entries[peripheralID] = entry
    }
    return (entry.lease, shouldCancel)
  }

  func finish(peripheralID: UUID) -> BluetoothConnectionLease? {
    entries.removeValue(forKey: peripheralID)?.lease
  }

  func reset() -> [BluetoothConnectionLease] {
    let leases = entries.values.map(\.lease)
    entries.removeAll()
    return leases
  }

  func activeLease(for peripheralID: UUID) -> BluetoothConnectionLease? {
    entries[peripheralID]?.lease
  }
}

File: desktop/windows/src/renderer/src/device/session/bluetoothConnectionLease.ts
/**
 * Connection attempt identity fencing — Windows port of macOS
 * Session/BluetoothConnectionLease.swift. A GATT connect cannot be reliably
 * cancelled mid-flight, so one attempt's identity stays pinned (the entry is
 * kept, merely marked cancelling) until a terminal event reports the attempt
 * dead; a later session can neither start on top of it nor consume its stale
 * completion.
 */

export interface BluetoothConnectionLease {
  deviceId: string
  token: number
  sessionGeneration: number
}

export class BluetoothConnectionLeaseError extends Error {
  readonly kind: 'leaseAlreadyActive' | 'centralUnavailable'

  private constructor(kind: 'leaseAlreadyActive' | 'centralUnavailable', message: string) {
    super(message)
    this.name = 'BluetoothConnectionLeaseError'
    this.kind = kind
  }

  static leaseAlreadyActive(): BluetoothConnectionLeaseError {
    return new BluetoothConnectionLeaseError(
      'leaseAlreadyActive',
      'A previous Bluetooth connection attempt is still draining'
    )
  }

  static centralUnavailable(): BluetoothConnectionLeaseError {
    return new BluetoothConnectionLeaseError('centralUnavailable', 'Bluetooth is not powered on')
  }
}

type LeasePhase = 'connecting' | 'connected' | 'cancelling'

interface LeaseEntry {
  lease: BluetoothConnectionLease
  phase: LeasePhase
}

export class BluetoothConnectionLeaseRegistry {
  private entries = new Map<string, LeaseEntry>()
  private nextToken = 1

  /** Starts a connection attempt. Throws while ANY entry exists for the
   *  device — including one that is only draining after cancellation. */
  begin(deviceId: string, sessionGeneration: number): BluetoothConnectionLease {
    if (this.entries.has(deviceId)) {
      throw BluetoothConnectionLeaseError.leaseAlreadyActive()
    }
    const lease: BluetoothConnectionLease = {
      deviceId,
      token: this.nextToken++,
      sessionGeneration
    }
    this.entries.set(deviceId, { lease, phase: 'connecting' })
    return lease
  }

  markConnected(lease: BluetoothConnectionLease): void {
    const entry = this.entries.get(lease.deviceId)
    if (entry === undefined || entry.lease.token !== lease.token) return
    if (entry.phase === 'connecting') entry.phase = 'connected'
  }

  /** Marks the attempt cancelling but KEEPS the entry — identity stays fenced
   *  until the platform reports a terminal event for it. */
  requestCancellation(lease: BluetoothConnectionLease): void {
    const entry = this.entries.get(lease.deviceId)
    if (entry === undefined || entry.lease.token !== lease.token) return
    entry.phase = 'cancelling'
  }

  /** Terminal event for the attempt: releases the fence. Stale tokens no-op. */
  end(lease: BluetoothConnectionLease): void {
    const entry = this.entries.get(lease.deviceId)
    if (entry === undefined || entry.lease.token !== lease.token) return
    this.entries.delete(lease.deviceId)
  }

  activeLease(deviceId: string): BluetoothConnectionLease | null {
    return this.entries.get(deviceId)?.lease ?? null
  }

  /** Central reset: every fence is void because the platform state is gone. */
  reset(): void {
    this.entries.clear()
  }
}
